docs/development.md

```diff
@@ -0,0 +1,829 @@
+indexed can change in the future.
+
+Using the integration configuration that was provided as a sample earlier, data
+will be written to disk in the following structure (relative to the
+integration's current working directory).
+
+To assist with debugging and visibilty into exactly what data was collected, the
+integration will bucket data collected from each step. Here is
+an example of what the `.j1-integration` directory may look like.
+
+```
+.j1-integration/
+  /index/
+    /entities/
+      _type
+        my_integration_account/
+          11fa25fb-dfbf-43b8-a6e1-017ad369fe98.json
+        my_integration_group/
+          f983f07d-f7d8-4f8e-87da-743940a5f48d.json
+          a76695f8-7d84-411e-a4e1-c012de041034.json
+        my_integration_user/
+          9cb7bee4-c037-4041-83b7-d532488f26a3.json
+          96992893-898d-4cda-8129-4695b0323642.json
+    /relationships
+      _type/
+        my_integration_user_to_group_relationship/
+          8fcc6865-817d-4952-ac53-8248b357b5d8.json
+  /graph
+    /step-fetch-accounts
+      /entities/
+        11fa25fb-dfbf-43b8-a6e1-017ad369fe98.json
+      /relationships
+    /step-fetch-users
+      /entities
+        9cb7bee4-c037-4041-83b7-d532488f26a3.json
+        96992893-898d-4cda-8129-4695b0323642.json
+      /relationships
+    /step-fetch-groups
+      /entities
+        a76695f8-7d84-411e-a4e1-c012de041034.json
+        f983f07d-f7d8-4f8e-87da-743940a5f48d.json
+      /relationships
+    /step-build-relationships
+        /relationships
+          8fcc6865-817d-4952-ac53-8248b357b5d8.json
+```
+
+Integration data that is staged for publishing will be stored under the
+`.j1-integration/graph` directory. Files containing Entities will be suffixed
+with `.entities.json` and files containing Relationships will be suffixed with
+`.relationships.json`. This is done because the directory structure of the
+`graph` directory is meant to assist with debugging and provide developrs
+insight about the data collect in each step. During synchronization, the
+directory will be blind
+
+Data will be indexed by `_type` and stored under the
+`.j1-integration/index/_type` directory.
+
+Each `json` file will store data under the following format (describe with a
+typescript interface):
+
+```
+interface FlushedEntityGraphData {
+  entities: Entity[]
+}
+
+interface FlushedRelationshipGraphData {
+  relationships: Relationship[]
+}
+```
+
+It's best to keep a "collect and forget" mindset to avoid retaining too much
+collected data in memory. Flushed data can always be pulled back into memory
+later via the `listEntitiesByType` and `listRelationshipsByType` and reading
+data from disk is decently fast.
+
+#### Failure handling
+
+By default, the framework will only halt the execution of an integration if the
+configuration validation fails. Failures that occur during the execution of a
+step will not halt the execution of later steps that depend on it.
```

Comment:
I was thinking that if step B depends on step A, and step A raises an error/fails, we would not run step B because it depends on step A. If we do execute step B anyway, it will have to find out if something went wrong in step A to avoid referencing data that may not exist, right?

Reply:
I was thinking that a failure in a previous step does not necessarily mean that there is an absence of data to work on. For example, if I am able to fetch users but only half of the groups, I should be able to still build relationships between the users and groups where possible. And since the relationship step relies on a step that failed, it's status would be marked as `partial_success_due_to_dependency_failure` and the types that it has produced would be included in the `partialDatasets` sent to the synchronizer.